sdks/java/extensions/sorter/src/main/java/org/apache/beam/sdk/extensions/sorter/ExternalSorter.java

```diff
@@ -0,0 +1,237 @@
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.UUID;
+import org.apache.beam.sdk.values.KV;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.BytesWritable;
+import org.apache.hadoop.io.SequenceFile;
+import org.apache.hadoop.io.SequenceFile.CompressionType;
+import org.apache.hadoop.io.SequenceFile.Sorter.RawKeyValueIterator;
+import org.apache.hadoop.io.SequenceFile.Writer;
+import org.apache.hadoop.mapred.JobConf;
+
+/** Does an external sort of the provided values using Hadoop's {@link SequenceFile}. */
+public class ExternalSorter implements Sorter {
+  private Options options;
+
+  /** Whether {@code sort()} was already called. */
+  private boolean sortCalled = false;
+
+  /** SequenceFile Writer for writing all input data to a file. */
+  private Writer writer;
+
+  /** Sorter used to sort the input file. */
+  private SequenceFile.Sorter sorter;
+
+  /** Temporary directory for input and intermediate files. */
+  private Path tempDir;
+
+  /** The list of input files to be sorted. */
+  private Path[] paths;
+
+  private boolean initialized = false;
+
+  /** {@code Options} contains configuration of the sorter. */
+  public static class Options implements Serializable {
+    private String tempLocation = "/tmp";
+    private int memoryMB = 100;
+
+    /** Sets the path to a temporary location where the sorter writes intermediate files. */
+    public void setTempLocation(String tempLocation) {
+      if (tempLocation.startsWith("gs://")) {
+        throw new IllegalArgumentException("Sorter doesn't support GCS temporary location.");
+      }
+
+      this.tempLocation = tempLocation;
+    }
+
+    /** Returns the configured temporary location. */
+    public String getTempLocation() {
+      return tempLocation;
+    }
+
+    /** Sets the size of the memory buffer in megabytes. */
+    public void setMemoryMB(int memoryMB) {
+      this.memoryMB = memoryMB;
+    }
+
+    /** Returns the configured size of the memory buffer. */
+    public int getMemoryMB() {
+      return memoryMB;
+    }
+  }
+
+  /** Returns a {@code Sorter} configured with the given {@code options}. */
+  public static ExternalSorter create(Options options) {
+    return new ExternalSorter(options);
+  }
+
+  /**
+   * Adds a given record to the sorter.
+   *
+   * <p>Records can only be added before calling {@code sort()}.
+   */
+  @Override
+  public void add(KV<byte[], byte[]> record) throws IOException {
+    initHadoopSorter();
+
+    if (sortCalled) {
+      throw new IllegalStateException("Records can only be added before sort().");
+    }
+
+    BytesWritable key = new BytesWritable();
+    BytesWritable value = new BytesWritable();
+    key.set(record.getKey(), 0, record.getKey().length);
+    value.set(record.getValue(), 0, record.getValue().length);
```

Comment:
Is there a reason not to use [`new BytesWritable(record.getKey())`](https://hadoop.apache.org/docs/r2.7.1/api/src-html/org/apache/hadoop/io/BytesWritable.html#line.53) and copying the array? Do you reuse the array / mutate it elsewhere?

Note to self: delete the comment if so.